include/revng/Pipeline/Step.h

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <memory>
+#include <set>
+#include <string>
+#include <type_traits>
+#include <utility>
+#include <vector>
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/ADT/iterator_range.h"
+#include "llvm/Support/Error.h"
+
+#include "revng/ADT/STLExtras.h"
+#include "revng/Pipeline/ContainerSet.h"
+#include "revng/Pipeline/Context.h"
+#include "revng/Pipeline/Contract.h"
+#include "revng/Pipeline/Pipe.h"
+#include "revng/Support/Debug.h"
+#include "revng/Support/FunctionIntrospection.h"
+
+namespace Pipeline {
+
+/// A step is a list of pipes that must be exectuted entirelly or not at all.
+/// Furthermore a step has a set of containers associated to it as well that
+/// will contain the element used for perform the computations.
+class Step {
+private:
+  std::string Name;
+  ContainerSet Containers;
+  std::vector<PipeWrapper> Pipes;
+
+public:
+  /// Clones the Targets from the backing containers of this step
+  /// and excutes all the pipes in sequence contained by this step
+  /// and returns the transformed containers.
+  ///
+  /// The contained values stays unchanged.
+  ContainerSet cloneAndRun(Context &Ctx,
+                           const ContainersStatus &Targets,
+                           llvm::raw_ostream *OS = nullptr);
+
+  /// Returns the set of goals that are already contained in the backing
+  /// containers of this step, futhermore adds to the container ToLoad those
+  /// that were not present.
+  ContainersStatus satisfiableGoals(const ContainersStatus &RequiredGoals,
+                                    ContainersStatus &ToLoad) const;
```

Comment:
Can you try to explain to me with an example how this method is supposed to be used?

Reply:
let us say that the state of the pipeline is 

serialization point A with targets 1a, 2a
step transform Xa in Xb
serialization point B with targets 1b

you ask satisfiableGoals how to satisfy the request 1b, 2b in the serialization point B

it will tell you with the returned container that 1b is already present and the container to Load will contain 2b so that one can propagate it backward to the previous step.

Reply:
Aah, no I see. However, this functions seems to do two completely unrelated things, 1) compute requirements and 2) tell you what's already available. I would consider splitting it in 2 (unless there's a real performance hit in doing that) and letting the caller do the rest.

Alternatively you could have something like:

```c++
struct RequirementsResult {
  ContainerStatus ImmediatelyAvailable;
  ContainerStatus MissingRequirements;
};
RequirementsResult computeRequirements(const ContainersStatus &RequiredGoals) const;
```

Reply:
yeah i can change the signature, but it is a performance hit to not do them together, they are very intertwined operations

Reply:
no i have to keep the signature, because i keep passing the same ToLoad container which gets populated for each step, it is not always empty at the start

Reply:
You could copy the result.

Otherwise
```c++
/// \p RequiredGoals ???
/// \p ImmediatelyAvailable ??? to be populated with ...
/// \p MissingRequirements ??? to be populated with ...
void computeRequirements(const ContainersStatus &RequiredGoals,
                         ContainerStatus &ImmediatelyAvailable
                         ContainerStatus &MissingRequirements) const;
```

Reply:
yeah copying the result would be the correct thing to do, but it is expensive when there are thousands of elements and this is used in the critical path of the loop.

the suggested signature looks better, but ImmediateAvailable is always empty at the start, because it does not make much sense for it not to be, and so would be best as return value.

Reply:
I've looked at the implementation and now it's a bit more clear.

```c++
/// \p RequiredGoals the required goals.
/// \p AlreadyAvailable out parameter where set of goals/targets that need no action will be added.
/// \return the targets required in input in order to produce \p RequiredGoals.
ContainerStatus analyzeGoals(const ContainersStatus &RequiredGoals,
                             ContainerStatus &AlreadyAvailable) const;
```

Make sure that avoding copy is not early optimization. The first thing you do in the implementation of this function is copying `RequiredGoals` and `AlreadyAvailable` AFAIU is a subset of that.